README.md
# BitStruct #

Class for packed binary data stored in ruby Strings. BitStruct accessors, generated from user declared fields, use pack/unpack to treat substrings as fields with a specified portable format.

Field types include:

* signed and unsigned integer (1..16 bits, or 24, 32, 40, 48... bits)

* numeric fields (signed, unsigned, float) can be designated as any of the following endians: little, big, native, network (default)

* fixed point, with arbitrary scale factor

* fixed length character array

* null-terminated character array for printable text

* octets (hex and decimal representation options; useful for IP and MAC addrs)

* float

* nested BitStruct

* vectors of embedded BitStructs

* free-form "rest" field (e.g., for the variable-size payload of a packet)

Field options (specifiable as :foo => val or "foo" => val) include:

* *display_name*: used in BitStruct#inspect_detailed and BitStruct#describe outputs.

* *default*: default field value

* *format*: alternate format string for inspect

* *endian*: for byte ordering of numeric fields (unsigned, signed, float): little, big, native, network (default)

* *fixed*: float stored as fixed-point integer, with specified scale factor


## Installation ##

For .gem:

    gem install bit-struct

For .tgz, unpack and then:

    ruby install.rb config
    ruby install.rb setup
    ruby install.rb install

## Uses ##

BitStruct is useful for defining packets used in network protocols. This is especially useful for raw IP--see examples/ping-recv.rb. All multibyte numeric fields are stored by default in network order.

BitStruct is most efficient when your data is primarily treated as a binary string, and only secondarily treated as a data structure. (For instance, you are routing packets from one socket to another, possibly looking at one or two fields as it passes through or munging some headers.) If accessor operations are a bottleneck, a better approach is to define a class that wraps an array and uses pack/unpack when the object needs to behave like a binary string.

## Features ##

* Extensible with user-defined field classes.

* Fields are fully introspectable and can be defined programmatically.

* BitStruct.describe prints out documentation of all the fields of a BitStruct subclass, based on declarations. This is useful for communicating with developers who are not using ruby, but need to talk the same protocols. See Example, below.

* Fields are inherited by subclasses. (The free-form "rest" field does not inherit, because it usually represents a payload whose structure is defined in subclasses using the fixed-size fields.)

* BitStruct#inspect and BitStruct#inspect_detailed can be used for prettified display of contents. (More generally, BitStruct#inspect takes some options that control formatting and detail level.) See Example, below.

* BitStruct inherits from String, so all the usual methods are available, and string-sharing (copy-on-write) is in effect.

* Easy access to a "prototype" instance of each BitStruct subclass, from which all instances of that subclass are initialized as a copy (in the absence of other initialization parameters, such as a hash, a string, or a block). See BitStruct.initial_value, and BitStruct#initialize. See Example, below.

* Easy conversion to and from hashes, using BitStruct#to_h and BitStruct.new.

* BitStructs can persist using Marshal (a BitStruct is after all just a string) or using YAML (with human readable representation of the fields).

* Includes tests, examples, and rdoc API documentation.

## Limitations ##

* Fields that are not aligned on byte boundaries may cross no more than two bytes boundaries. (See examples/byte-bdy.rb.)

* No variable length fields (except the #rest field).
 
## Future plans ##

* Currently, the library is written in pure ruby. The implementation uses Array#pack and String#unpack calls, as well as shifting and masking in pure ruby. Future versions will optionally generate a customized C extension for better efficiency.

* A debug mode in which a class identifier is prepended to every BitStruct, so that protocol errors can be detected. (This feature has been implemented in an app that uses BitStruct, but needs to be refactored into the BitStruct library itself.)

* Remove field size and alignment limitations.

## Example ##

An IP packet can be defined and used like this:

```ruby
    require 'bit-struct'

    class IP < BitStruct
      unsigned    :ip_v,     4,     "Version"
      unsigned    :ip_hl,    4,     "Header length"
      unsigned    :ip_tos,   8,     "TOS"
      unsigned    :ip_len,  16,     "Length"
      unsigned    :ip_id,   16,     "ID"
      unsigned    :ip_off,  16,     "Frag offset"
      unsigned    :ip_ttl,   8,     "TTL"
      unsigned    :ip_p,     8,     "Protocol"
      unsigned    :ip_sum,  16,     "Checksum"
      octets      :ip_src,  32,     "Source addr"
      octets      :ip_dst,  32,     "Dest addr"
      rest        :body,            "Body of message"

      note "     rest is application defined message body"

      initial_value.ip_v    = 4
      initial_value.ip_hl   = 5
    end

    ip = IP.new
    ip.ip_tos = 0
    ip.ip_len = 0
    ip.ip_id  = 0
    ip.ip_off = 0
    ip.ip_ttl = 255
    ip.ip_p   = 255
    ip.ip_sum = 0
    ip.ip_src = "192.168.1.4"
    ip.ip_dst = "192.168.1.255"
    ip.body   = "This is the payload text."
    ip.ip_len = ip.length

    puts ip.inspect
    puts "-"*50
    puts ip.inspect_detailed
    puts "-"*50
    puts IP.describe
```

(Note that you can also construct an IP packet by passing a string to new, or by passing a hash of <tt>field,value</tt> pairs, or by providing a block that is yielded the new BitStruct.)

The output of this fragment is:

    #<IP ip_v=4, ip_hl=5, ip_tos=0, ip_len=45, ip_id=0, ip_off=0, ip_ttl=255, ip_p=255, ip_sum=0, ip_src="192.168.1.4", ip_dst="192.168.1.255", body="This is the payload text.">
    --------------------------------------------------
    IP:
                           Version = 4
                     Header length = 5
                               TOS = 0
                            Length = 45
                                ID = 0
                       Frag offset = 0
                               TTL = 255
                          Protocol = 255
                          Checksum = 0
                       Source addr = "192.168.1.4"
                         Dest addr = "192.168.1.255"
                   Body of message = "This is the payload text."
    --------------------------------------------------

    Description of IP Packet:
        byte: type         name          [size] description
    ----------------------------------------------------------------------
          @0: unsigned     ip_v          [  4b] Version
          @0: unsigned     ip_hl         [  4b] Header length
          @1: unsigned     ip_tos        [  8b] TOS
          @2: unsigned     ip_len        [ 16b] Length
          @4: unsigned     ip_id         [ 16b] ID
          @6: unsigned     ip_off        [ 16b] Frag offset
          @8: unsigned     ip_ttl        [  8b] TTL
          @9: unsigned     ip_p          [  8b] Protocol
         @10: unsigned     ip_sum        [ 16b] Checksum
         @12: octets       ip_src        [ 32b] Source addr
         @16: octets       ip_dst        [ 32b] Dest addr
         rest is application defined message body

## Web site ##

The current version of this software can be found at https://github.com/vjoel/bit-struct.

## License ##

This software is distributed under the Ruby license. See http://www.ruby-lang.org.

## Author ##

Joel VanderWerf, mailto:vjoel@users.sourceforge.net
Copyright (c) 2005-2014, Joel VanderWerf.


History.txt
bit-struct 0.15

- modernize rakefile and gemspec
- switch to minitest

bit-struct 0.14

- updated for ruby 2.0

bit-struct 0.13.6

- Fixed bug in #to_a. Thanks Michael Edgar.

bit-struct 0.13

- Packaged as a gem using bones.

- Added nest fields defined with blocks. (Markus Fischer)

- Vector length may be specified as argument or :length option. (Markus Fischer)

bit-struct 0.12

- Added vectors.

- BitStruct#initialize can take an IO argument.

bit-struct 0.11

- Allow unaligned fields to cross up to two byte boundaries.
  See examples/byte-bdy.rb.

bit-struct 0.10

- Fixed a bug when calling #to_yaml on a BitStruct with a pad field. Thanks
  to Jay Reitz for reporting it and providing the solution.

- Added BitStruct.default_options. Particularly useful for default endian
  settings--see examples/native.rb.

- Fixed a bug that prevented warning about field name conflicts with
  existing methods when the name is given as a symbol.

- Fixed point fields may now have fractional divisors (like :fixed => 0.001).

bit-struct 0.9

- Added examples/field-ripper.rb.

- Added BitStruct#field_by_name.

- Added more warnings about nested field accessors returning a *copy* of the
  nested data.

- Added "pad" fields. See documentation in pad-field.rb and examples/pad.rb.

- The #initial_value method now yields the value to a block, if given.

- A BitStruct class is only closed when an instance is created. (After this
  point no fields can be added.) Formerly, a class was closed when any method
  requested the total length of the structure. This change makes it easier
  to add groups of fields modularly, each with their own initial_value block.

- Added examples/modular-def.rb to explain how to factor a BitStruct
  definition into modules.

bit-struct 0.8

- Signed fields can now (like unsigned fields) be any multiple of 8 bits, and are accessed as fixnums or bignums, as needed.

- It's easier to subclass BitStruct::OctetField. BitStruct::HexOctetField is now implemented in this way, just by defining three constants and two trivial class methods.

bit-struct 0.7

- BitStruct.describe now takes an option hash, and the :expand=>true option causes nested fields to be expanded in the description.

- Unsigned integer fields can now be any multiple of 8 bytes: 8, 16, 24, 32, 40, ... (Fields of 1..15 bits in length are of course still supported.)

- Added the :endian => :little option to signed integer, unsigned integer, and float fields. The full set of endian options is now [:little, :big, :network, :native]. The default is always :network.

- Option names may be strings or symbols. Values that can be symbols can also be strings.

- Added examples/bignum.rb.

- Added support for the YAML in ruby 1.8.2 (the YAML in 1.8.4 was already supported).

bit-struct 0.6

- Added the :endian => :native option for numerical fields (signed, unsigned, float).

- Fixed error message with 9..15 bit fields aligned on byte boundary.

- The #initial_value is now inherited (before applying defaults).

- New examples: raw.rb, native.rb.

bit-struct 0.5

- Integer fields may now cross byte boundaries, as long as the field fits within two whole bytes.

bit-struct 0.4

- Fixed a bug in reading text and char fields in YAML: if the value was interpreted by YAML as something other than a string, an error would result.

- When BitStructs are loaded from yaml, the key is treated as a setter method, rather than a field. This is useful in case there is a field that needs special setters to accept humanly readable input. (For example, a char field with length-prefixed subfields.)

bit-struct 0.3

- BitStruct classes are now YAML friendly.

- The default behavior of BitStruct#inspect and BitStruct#inspect_detailed is changed to print out the "rest" field, if any. This can be disabled by passing inspect an options hash with :include_rest => false. See BitStruct::DEFAULT_INSPECT_OPTS and BitStruct::DETAILED_INSPECT_OPTS. See examples/ip.rb and examples/rest.rb.

- The default behavior of BitStruct#to_h is changed to include the "rest" field. As above, this can be disabled by passing to_h an options hash with :include_rest => false.

- The default behavior of BitStruct#to_a is changed to include the "rest" field. As above, this can be disabled by passing false as an argument to to_a.

bit-struct 0.2

- first public release

bit-struct 0.1

- first release


examples/ip.rb
require 'bit-struct'

class IP < BitStruct
  unsigned    :ip_v,     4,     "Version"
  unsigned    :ip_hl,    4,     "Header length"
  unsigned    :ip_tos,   8,     "TOS"
  unsigned    :ip_len,  16,     "Length"
  unsigned    :ip_id,   16,     "ID"
  unsigned    :ip_off,  16,     "Frag offset"
  unsigned    :ip_ttl,   8,     "TTL"
  unsigned    :ip_p,     8,     "Protocol"
  unsigned    :ip_sum,  16,     "Checksum"
  octets      :ip_src,  32,     "Source addr"
  octets      :ip_dst,  32,     "Dest addr"
  rest        :body,            "Body of message"

  note "     rest is application defined message body"
  
  initial_value.ip_v    = 4
  initial_value.ip_hl   = 5
end

if __FILE__ == $0
  ip1 = IP.new
  ip1.ip_tos = 0
  ip1.ip_len = 0
  ip1.ip_id  = 0
  ip1.ip_off = 0
  ip1.ip_ttl = 255
  ip1.ip_p   = 255
  ip1.ip_sum = 0
  ip1.ip_src = "192.168.1.4"
  ip1.ip_dst = "192.168.1.255"
  ip1.body   = "This is the payload text."
  ip1.ip_len = ip1.length
  
  ip2 = IP.new do |ip|
    ip.ip_tos = 0
    ip.ip_len = 0
    ip.ip_id  = 0
    ip.ip_off = 0
    ip.ip_ttl = 255
    ip.ip_p   = 255
    ip.ip_sum = 0
    ip.ip_src = "192.168.1.4"
    ip.ip_dst = "192.168.1.255"
    ip.body   = "This is the payload text."
    ip.ip_len = ip.length
  end
  
  ip3 = IP.new(
    :ip_tos => 0,
    :ip_len => 0,
    :ip_id  => 0,
    :ip_off => 0,
    :ip_ttl => 255,
    :ip_p   => 255,
    :ip_sum => 0,
    :ip_src => "192.168.1.4",
    :ip_dst => "192.168.1.255",
    :body   => "This is the payload text."
  ) do |ip|
    ip.ip_len = ip.length
  end
  
  ip4 = IP.new(ip1) # Construct from a BitStruct (or String)
  
  raise unless ip1 == ip2
  raise unless ip1 == ip3
  raise unless ip1 == ip4

  ip = ip1
  
  puts ip.inspect
  puts "-"*50
  puts ip.inspect_detailed
  puts "-"*50
  puts
  puts "Description of IP Packet:"
  puts IP.describe
end
